### PR TITLE
Deploy to Kubernetes section

### PR DESCRIPTION
**Pull request by Envek**
How to deploy anycable with Helm to Kubernetes cluster.

This manual consists of two sections:
 1. How to deploy anycable-go. It is easy: just use our chart!
 2. How to deploy one more instance of your application to run as anycable-rpc. That is harder because it is hard to extract to chart or anything as all ruby applications are different. So, only examples of k8s resources to be converted into helm templates manually.

Most probably we need to add more introduction wording like “why we do it in that way.” WDYT?

### CODE REVIEW

File: docs/deployment/kubernetes.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
+# Kubernetes Deployment
+
+## AnyCable-Go
+
+anycable-go can be easily deployed to your Kubernetes cluster using Helm and [our official Helm chart][anycable-helm].
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]:
```suggestion
AnyCable-Go can be easily deployed to your Kubernetes cluster using Helm and [our official Helm chart][anycable-helm].
```

File: docs/deployment/kubernetes.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+ 1. Add it as a dependency to your main application:
+
+    ```yaml
+    # Chart.yaml
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]: It's important to mention that it makes sense for Helm 3 only.

File: docs/deployment/kubernetes.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    dependencies:
+    - name: anycable-go
+      version: 0.2.4
+      repository: https://helm.anycable.io/
+    ```
+
+    Check latest Helm chart version at [github.com/anycable/anycable-helm/releases](https://github.com/anycable/anycable-helm/releases).
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]:
```suggestion
    Check the latest Helm chart version at [github.com/anycable/anycable-helm/releases](https://github.com/anycable/anycable-helm/releases).
```

File: docs/deployment/kubernetes.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        anycableRedisUrl: redis://:password@redis-anycable:6379/0
+      ingress:
+        acme: # if you're using Let's Encrypt
+          hosts:
+            - your-app.com
+    ```
+
+Read the [chart’s README][anycable-helm] for more info.
+
+## Ruby RPC server
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]: I guess `Ruby RPC server` is not a convenient name. Everywhere `RPC server` is mentioned.

File: docs/deployment/kubernetes.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    metadata:
+      name: anycable-rpc
+      labels:
+        component: anycable-rpc
+    spec:
+      selector:
+        component: anycable-rpc
+      type: ClusterIP
+      ports:
+        - port: 50051
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md:114]: Should we extract port number to values?

File: docs/deployment/kubernetes.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    # Chart.yaml
+    dependencies:
+    - name: anycable-go
+      version: 0.2.4
+      repository: https://helm.anycable.io/
+    ```
+
+    Check latest Helm chart version at [github.com/anycable/anycable-helm/releases](https://github.com/anycable/anycable-helm/releases).
+
+ 1. And then configure it in your application values within `anycable-go` namespace:
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]: What namespace is meant here? Should it be a subdirectory?

File: docs/deployment/kubernetes.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        anycableRpcHost: anycable-rpc:50051
+      ingress:
+        enable: true
+        path: /cable
+
+    # values/production.yaml
+    anycable-go:
+      env:
+        # Assuming that Redis is available in K8s in the same namespace as redis-anycable service
+        anycableRedisUrl: redis://:password@redis-anycable:6379/0
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]: I'm not sure that it's a good practice even as an example. Peeps will be copypasting it.

File: docs/deployment/kubernetes.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        acme: # if you're using Let's Encrypt
+          hosts:
+            - your-app.com
+    ```
+
+Read the [chart’s README][anycable-helm] for more info.
+
+## Ruby RPC server
+
+To run Ruby counterpart of Anycable which will handle connection authentication and execute your business logic we will need to create separate deployment and service for it.
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]:
```suggestion
To run Ruby counterpart of AnyCable which will handle connection authentication and execute your business logic we need to create a separate deployment and a corresponding service for it.
```

File: docs/deployment/kubernetes.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+              image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+              imagePullPolicy: IfNotPresent
+              command:
+                - bundle
+                - exec
+                - anycable
+                - !!str --rpc-host=0.0.0.0:50051
+              env:
+                - name: ANYCABLE_REDIS_URL
+                  value: "redis://:password@redis-anycable:6379/0" # Same as for anycable-go
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]: Ditto

File: docs/deployment/kubernetes.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+          containers:
+            - name: anycable-rpc
+              image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+              imagePullPolicy: IfNotPresent
+              command:
+                - bundle
+                - exec
+                - anycable
+                - !!str --rpc-host=0.0.0.0:50051
+              env:
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md:90]: Should we mention to include the rest of the env requred by the application itself?

File: docs/deployment/kubernetes.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                - name: ANYCABLE_REDIS_URL
+                  value: "redis://:password@redis-anycable:6379/0" # Same as for anycable-go
+                  # or, better, place it into secrets for better security:
+                  # valueFrom:
+                  #   secretKeyRef:
+                  #     name: "anycable-go-secrets"
+                  #     key: anycableRedisUrl
+              resources:
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]: Does it make sense to extract resources definition to the values?

File: docs/deployment/kubernetes.md (hunk 6 above)
Comment (Envek) [reply docs/deployment/kubernetes.md]: Here "namespace" is a section in yaml file with values. It must be named `anycable-go` (equals to the chart name)

File: docs/deployment/kubernetes.md (hunk 5 above)
Comment (Envek) [reply docs/deployment/kubernetes.md:114]: Yes, but it will make examples way harder to understand

File: docs/deployment/kubernetes.md (hunk 10 above)
Comment (Envek) [reply docs/deployment/kubernetes.md:90]: Yes 👍

File: docs/deployment/kubernetes.md (hunk 5 above)
Comment (palkan) [reply docs/deployment/kubernetes.md:114]: Let's add a comment like this:

```
# you should define these parameters in the values.yml file, we give them here directly for readability
```

File: docs/deployment/kubernetes.md (hunk 2 above)
Comment (Envek) [reply docs/deployment/kubernetes.md]: Added note

File: docs/deployment/kubernetes.md (hunk 4 above)
Comment (Envek) [reply docs/deployment/kubernetes.md]: Fixed

File: docs/deployment/kubernetes.md (hunk 11 above)
Comment (Envek) [reply docs/deployment/kubernetes.md]: I decided to remove it altogether – this example is already too long

**Comment by Envek** [comment]
Done with your comments. Anything else?

File: docs/deployment/kubernetes.md [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      ingress:
+        enable: true
+        path: /cable
+
+    # values/production.yaml
+    anycable-go:
+      env:
+        # Assuming that Redis is available in K8s in the same namespace as redis-anycable service
+        anycableRedisUrl: redis://:CHANGE-THE-PASSWORD@redis-anycable:6379/0
+        anycableRpcHost: "anycable-rpc:50051" # See service definition for RPC server
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]: We already have this line in `values.yaml`
```suggestion
```

File: docs/deployment/kubernetes.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+            - your-app.com
+    ```
+
+Read the [chart’s README][anycable-helm] for more info.
+
+## RPC server
+
+To run Ruby counterpart of AnyCable which will handle connection authentication and execute your business logic we need to create a separate deployment and a corresponding service for it.
+
+ 1. [**Deployment**](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) that will spin up required number of pods and handle rolling restarts on deploys
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]:
```suggestion
 1. [**Deployment**](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) that will spin up a required number of pods and handle rolling restarts on deploys
```

File: docs/deployment/kubernetes.md [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        component: anycable-rpc
+      type: ClusterIP
+      ports:
+        # you should define these parameters in the values.yml file, we give them here directly for readability
+        - port: 50051
+          targetPort: 50051
+          protocol: TCP
+    ```
+
+ 1. (Optional) [**network policy**](https://kubernetes.io/docs/concepts/services-networking/network-policies/) will restrict access to pods running RPC service to only those that runs anycable-go daemon in the same namespace.
```
Comment (sponomarev) [inline docs/deployment/kubernetes.md]:
```suggestion
 1. (Optional) [**network policy**](https://kubernetes.io/docs/concepts/services-networking/network-policies/) will restrict access to pods running RPC service to only those that run AnyCable-Go daemon in the same namespace.
```

**Review by sponomarev — APPROVED** [review]
Just a couple of last suggestions and we're good to go

File: docs/deployment/kubernetes.md (hunk 12 above)
Comment (Envek) [reply docs/deployment/kubernetes.md]: Good catch, thank you!

**Review by palkan — APPROVED** [review]
👍

**Pull request merged**